### ISSUE TITLE
ssh connection troubleshooting

### ISSUE THREAD

**Issue by alexandrejbr** [erlang/otp#8718]
**Is your feature request related to a problem? Please describe.**
For troubleshooting purposes, we would like to be able to have more detailed information of setup of the an SSH connection. For instance the proposed algorithms by each role and the authentication methods. This would be useful for both connections that are established successfully and for connections that fail to be established.

What we have in mind is to be able to debug a connection like one would do in openssh with the verbose modes, but since that is perhaps quite ambitious, starting with what each role is proposes for the connection and information about the authentications attempts would already be great.
**Describe the solution you'd like**
It's hard to imagine what would be the best way to obtain this information, but I imagine that a callback could work for both roles, even though for the client role the error would come from the result of ssh:connect and then a callback function would be called as well.

Logging also works, but perhaps a bit less flexible.

Do you think it would be interesting to have such troubleshooting capabilities in the ssh application?

**Comment by u3s** [comment]
have you seen `ssh_dbg` module? it is pretty powerful.

for seeing result of algorithm negotiation I would use:
`ssh:start(), ssh_dbg:on([alg]).`

and for also getting SSH messages leading to negotiation result:
`ssh:start(), ssh_dbg:on([alg, ssh_messages]).`

apply above before establishing connection. should work for both roles.
it is based on tracing feature.

**Comment by u3s** [comment]
you disable it with `ssh_dbg:off`.

it might be slightly challenging to predict which connection will fail ... traces and production systems are tricky to combine I'm afraid.

ssh_dbg is not documented feature so you would need to read source to understand how to it works.

**Comment by u3s** [comment]
ping

**Comment by u3s** [comment]
> If one could get similar information in callbacks or some other stream of events it would be great.

There was a discussion about some  generic solution related to tracing of OTP apps. Maybe even an OTP behavior.
Unfortunately they're having lower backlog priority and no details were agreed at this stage.

I guess your thoughts lean towards a possibility for subscribing to troubleshooting events, which you could then process in whatever way you like or store it for later.

**Comment by alexandrejbr** [comment]
> I guess your thoughts lean towards a possibility for subscribing to troubleshooting events, which you could then process in whatever way you like or store it for later.

Yes, I think you phrased it quite well. I think this could replace the existing logs and if one prefers to have the logs instead we could have a built in subscriber of these events that would translate them into info level logs.

**Comment by u3s** [comment]
What about this callbacks already present?

https://www.erlang.org/doc/apps/ssh/ssh.html#t:callbacks_daemon_options/0

**Comment by u3s** [comment]
ping?

**Comment by alexandrejbr** [comment]
@u3s I have not being so active because I wanted to come up with a more concrete proposal, but maybe we can keep a bit more high level discussing the problems before we reach that state.

We use both ssh client and server and build on top of it, so our use case of debugging an ssh connection is not that we are going to open an erlang shell to do the debugging. It's more like problems occurred and we want to look at what why after the fact.

I can give you one example of one situation where the error return lost the debug information and it's only to logs that it goes, which is in the ssh_transport:handle_kexinit_msg. In the client role the user code gets just the default text for the error.

Probably if the user could supply a custom log function and we would place the error code and the extra information in a proplist we would already be extremely good as far as error scenarios.

Then in cases all is good with a connection there may be some interest in knowing to know what was actually negotiated. For the client role one can use the ssh:connection_info function, for the server role there's way unless we use tracing or the ssh_dbg module, isn't that right?

In a nutshell I think one part of this feature request is a programatic way of customising the logs in case of error and other part is to be able to do some introspection in connections that were successfully established, maybe the connectfun (daemon callbacks) could have the connection "handler"/pid so we can call connection_info on it or just have the connection information there as a parameter of the callback.

Do you think any of these are something you would consider having?

**Comment by u3s** [comment]
> Do you think any of these are something you would consider having?

Your argumentation sounds convincing to me.
Unfortunately this has to wait until I have more time to investigate it further.

**Comment by alexandrejbr** [comment]
@u3s would it help if we do a draft PR? Something incomplete, but just enough to make it easier for you reflect about it.

**Comment by u3s** [comment]
sure. if you have a concept code, please share.

**Comment by alexandrejbr** [comment]
@u3s I have opened one draft PR that deals with the part of my suggestions. I will be opening another PR with the remaining suggestions.

Since they are kind of separate I thought it would be better to keep them separate so if you are only interested in one of the PRs then it's easier for us to move forward.

**Comment by alexandrejbr** [comment]
@u3s I have decided to use the same branch so you'll find everything in #9157. It would be great if you could have a look at it and check if this is something you would consider having. Not necessarily in the current form, I'm more than happy to refine it further.

**Comment by alexandrejbr** [comment]
@u3s we had some discussions internally and while the changes proposed in https://github.com/erlang/otp/pull/9157 will help us in almost all the cases there's still use cases where it would not be enough.

The cases we had in mind are:
- Detect that a message was not sent;
- A message was sent at the wrong time (symptom of a bug);
- A problem occurs after receiving a certain message;
- Prove that the SSH implementation based on the erlang ssh application is doing everything correct and that the problem is the non-erlang peer that is doing something not expected.

Currently we have the `unexpectedfun`. Would you consider adding a new callback that would give the opportunity for the user/developer to log the messages exchanged? I am thinking that the message type would be a good start.

So it could be something like

```erlang
-type messagefun_common_option() :: messsagefun(Direction::sent | received, Type:: integer(),
                                                                                           ExtraInfo::#{connection_info := Info::proplists:proplist()}) -> any()).
```

The idea of having the ExtraInfo map is to be able to extend the callback easier with some useful information like for instance receiver channel/sender channel.

**Comment by u3s** [comment]
sorry for delays, but I was swamped with other work. hope to get back to this one soonish.

**Comment by u3s** [comment]
we have other PR which replaces error_logger with logger 1st.
Then we will revisit this PR.

**Comment by u3s** [comment]
I would propose closing #9157 in favor of #11033 and #11034.
@alexandrejbr what do you think?
